File: proposed-projects/round-trip-experiment/reverse/20260909T185402Z_14873f98/20260910T102153Z_a4bbfe64/ported_implementation/src/grammarParser/buildRuleStack.ts
import { RuleRef } from '../grammarGraph/ruleRef';
import { isRange, isRuleEnd } from '../grammarGraph/typeGuards';
import { RuleChar, RuleCharExclude, RuleEnd } from '../grammarGraph/types';
import type { Range, UnresolvedRule } from '../grammarGraph/types';
import {
  isRuleDefAlt,
  isRuleDefChar,
  isRuleDefCharAlt,
  isRuleDefCharNot,
  isRuleDefCharRngUpper,
  isRuleDefEnd,
  isRuleDefRef,
} from '../rulesBuilder/typeGuards';
import type {
  InternalRuleDef,
  InternalRuleDefChar,
  InternalRuleDefCharNot,
} from '../rulesBuilder/types';

const makeCharRule = (
  ruleDef: InternalRuleDefChar | InternalRuleDefCharNot,
): RuleChar | RuleCharExclude =>
  isRuleDefCharNot(ruleDef)
    ? new RuleCharExclude([...ruleDef.value])
    : new RuleChar([...ruleDef.value]);

export const buildRuleStack = (
  linearRules: InternalRuleDef[],
): UnresolvedRule[][] => {
  const stack: UnresolvedRule[][] = [];
  let paths: UnresolvedRule[] = [];

  let idx = 0;
  while (idx < linearRules.length) {
    const ruleDef = linearRules[idx];
    if (isRuleDefChar(ruleDef) || isRuleDefCharNot(ruleDef)) {
      // this could be a single char, or a range, or a sequence of alts;
      // we don't know until we step through it.
      const charRule = makeCharRule(ruleDef);
      idx += 1;
      let rule: InternalRuleDef | undefined = linearRules[idx];
      while (
        idx < linearRules.length &&
        (isRuleDefCharRngUpper(rule) || isRuleDefCharAlt(rule))
      ) {
        if (isRuleDefCharRngUpper(rule)) {
          // previous rule value should be a number
          const prevValue = charRule.value.pop();
          if (isRange(prevValue)) {
            throw new Error(
              `Unexpected range, expected a number but got an array: ${JSON.stringify(prevValue)}`,
            );
          }
          if (prevValue === undefined) {
            throw new Error('Unexpected undefined value');
          }
          charRule.value.push([prevValue, rule.value] as Range);
        }
        if (isRuleDefCharAlt(rule)) {
          charRule.value.push(rule.value);
        }
        idx += 1;
        rule = linearRules[idx];
      }
      paths.push(charRule);
    } else {
      if (isRuleDefAlt(ruleDef)) {
        if (!paths.length) {
          throw new Error('Encountered alt without anything before it');
        }
        paths.push(new RuleEnd());
        stack.push(paths);
        paths = [];
      } else if (isRuleDefEnd(ruleDef)) {
        paths.push(new RuleEnd());
      } else if (isRuleDefRef(ruleDef)) {
        paths.push(new RuleRef(ruleDef.value));
      } else if (isRuleDefCharAlt(ruleDef)) {
        throw new Error(
          `Encountered char alt, should be handled by above block: ${ruleDef.type}`,
        );
      } else {
        throw new Error(`Unsupported rule type: ${ruleDef.type}`);
      }
      idx += 1;
    }
  }

  if (!(paths.length && isRuleEnd(paths[paths.length - 1]))) {
    paths.push(new RuleEnd());
  }

  stack.push(paths);

  return stack;
};

File: packages/gbnf-experiment/data/20260909T013603Z_054a84ca/ported_implementation/gbnf/grammar_parser/build_rule_stack.py
from __future__ import annotations

from typing import List, Sequence, Union

from ..grammar_graph.rule_ref import RuleRef
from ..grammar_graph.type_guards import is_range, is_rule_end
from ..grammar_graph.types import RuleChar, RuleCharExclude, RuleEnd
from ..rules_builder.type_guards import (
    is_rule_def_alt,
    is_rule_def_char,
    is_rule_def_char_alt,
    is_rule_def_char_not,
    is_rule_def_char_rng_upper,
    is_rule_def_end,
    is_rule_def_ref,
)
from ..rules_builder.types import InternalRuleDef
from ..utils.js import at

UnresolvedRule = Union[RuleChar, RuleCharExclude, RuleEnd, RuleRef]


def make_char_rule(rule_def: InternalRuleDef) -> Union[RuleChar, RuleCharExclude]:
    cls = RuleCharExclude if is_rule_def_char_not(rule_def) else RuleChar
    return cls(list(rule_def.value))


def build_rule_stack(linear_rules: Sequence[InternalRuleDef]) -> List[List[UnresolvedRule]]:
    paths: List[UnresolvedRule] = []

    stack: List[List[UnresolvedRule]] = []

    idx = 0
    while idx < len(linear_rules):
        rule_def = linear_rules[idx]
        if is_rule_def_char(rule_def) or is_rule_def_char_not(rule_def):
            # this could be a single char, or a range, or a sequence of alts; we don't
            # know until we step through it.
            char_rule = make_char_rule(rule_def)
            idx += 1
            rule = at(linear_rules, idx)
            while idx < len(linear_rules) and (
                is_rule_def_char_rng_upper(rule) or is_rule_def_char_alt(rule)
            ):
                if is_rule_def_char_rng_upper(rule):
                    # previous rule value should be a number
                    prev_value = char_rule.value.pop() if char_rule.value else None
                    if is_range(prev_value):
                        raise ValueError(
                            f'Unexpected range, expected a number but got an array: {prev_value}'
                        )
                    if prev_value is None:
                        raise ValueError('Unexpected undefined value')
                    char_rule.value.append([prev_value, rule.value])
                if is_rule_def_char_alt(rule):
                    char_rule.value.append(rule.value)
                idx += 1
                rule = at(linear_rules, idx)
            paths.append(char_rule)
        else:
            if is_rule_def_alt(rule_def):
                if not paths:
                    raise ValueError('Encountered alt without anything before it')
                paths.append(RuleEnd())
                stack.append(paths)
                paths = []
            elif is_rule_def_end(rule_def):
                paths.append(RuleEnd())
            elif is_rule_def_ref(rule_def):
                paths.append(RuleRef(rule_def.value))
            elif is_rule_def_char_alt(rule_def):
                raise ValueError(
                    f'Encountered char alt, should be handled by above block: {rule_def.type}'
                )
            else:
                raise ValueError(f'Unsupported rule type: {rule_def.type}')
            idx += 1

    if not is_rule_end(at(paths, len(paths) - 1)):
        paths.append(RuleEnd())

    stack.append(paths)

    return stack
